packages/component-aws-s3/README.md

```diff
@@ -0,0 +1,227 @@
+    }
+```
+
+## Consistently Maintaining a Database Table of S3 Objects
+In many cases, an application wants to keep track of objects that have been stored in S3.  S3 is, as the name implies, simple storage, and it doesn't track file attributes, permissions, fine-grained ownership, dependencies, etc.
+
+Keeping an indexed set of file metadata records, including referential links to their owners, is a "database problem".  And, while keeping the database in sync with the contents of S3 sounds like it may be tricky, [DBOS Transact Workflows](https://docs.dbos.dev/tutorials/workflow-tutorial) provide the perfect tool for accomplishing this, even in the face of client or server failures.
+
+The `S3Ops` class provides workflows that can be used to ensure that a table of file records is maintained for an S3 bucket.  This table can have any schema suitable to the application (an example table schema can be found in `s3_utils.test.ts`), because the application provides the code to maintain it as a set of callback functions that will be triggered from the workflow.
+
+The interface for the workflow functions (described below) allows for the following callbacks:
+```typescript
+export interface FileRecord {
+    key: string; // AWS S3 Key
+}
+
+export interface S3Config{
+    s3Callbacks?: {
+        /* Called when a new active file is added to S3 */
+        newActiveFile: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+
+        /* Called when a new pending (still undergoing workflow processing) file is added to S3 */
+        newPendingFile: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+
+        /* Called when pending file becomes active */
+        fileActivated: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+
+        /* Called when a file is deleted from S3 */
+        fileDeleted: (ctx: WorkflowContext, rec: FileRecord) => Promise<unknown>;
+    },
+    //... remainder of S3 Config
+}
+```
+
+### Workflow to Upload a String to S3
+The `saveStringToFile` workflow stores a string to an S3 key, and runs the callback function to update the database.  If anything goes wrong during the workflow, S3 will be cleaned up and the database will be unchanged by the workflow.
+
+```typescript
+    await ctx.invokeWorkflow(defaultS3).saveStringToFile(fileDBRecord, 'This is my file');
+```
+
+This workflow performs the following actions:
+* Puts the string in S3
+* If there is difficulty with S3, ensures that no entry is left there and throws an error
+* Invokes the callback for a new active file record
+
+***Note:*** **Function arguments to workflows, transactions, and communicators are logged to the database.  This means that workflow functions are suitable for writing kilobytes of data to S3, that megabytes will possibly work, but gigabytes will cause great pain.  Consider [having the client upload to S3 with a presigned post](#workflow-to-allow-client-file-upload) if the data is large.**
```

Comment:
We need to have this note somewhere, but maybe not four times?

Reply:
Qian wanted more of them, so I added more :-D.  FWIW the suggestions differ.

Reply:
I can tone them down, particularly the repetitive part?